File: magic.py
import random
BASIC_POWERS=["fire","metal","wood","earth","water"]
ALIGNMENTS=["light","dark"]
COSMIC_POWERS=["space","time"]
MOVES={
    "flame_burst":{"name":"Flame Burst","damage":10,"cooldown":0,"type":"fire","level":"basic"},
    "fireball":{"name":"Fireball","damage":13,"cooldown":1,"type":"fire","level":"basic"},
    "heat_wave":{"name":"Heat Wave","damage":17,"cooldown":1,"type":"fire","level":"basic"},
    "iron_spike":{"name":"Iron Spike","damage":10,"cooldown":0,"type":"metal","level":"basic"},
    "shard_spray":{"name":"Shard Spray","damage":15,"cooldown":3,"type":"metal","level":"basic"},
    "knife_storm":{"name":"Knife Storm","damage":15,"cooldown":3,"type":"metal","level":"basic"},
    "splinter":{"name":"Splinter","damage":10,"cooldown":0,"type":"wood","level":"basic"},
    "vine_whip":{"name":"Vine Whip","damage":12,"cooldown":1,"type":"wood","level":"basic"},
    "leaf_storm":{"name":"Leaf Storm","damage":15,"cooldown":2,"type":"wood","level":"basic"},
    "pebble_shot":{"name":"Pebble Shot","damage":10,"cooldown":0,"type":"earth","level":"basic"},
    "tremor":{"name":"Tremor","damage":15,"cooldown":1,"type":"earth","level":"basic"},
    "boulder_crush":{"name":"Boulder Crush","damage":20,"cooldown":4,"type":"earth","level":"basic"},
    "water_jet":{"name":"Water Jet","damage":10,"cooldown":0,"type":"water","level":"basic"},
    "tidal_wave":{"name":"Tidal Wave","damage":15,"cooldown":1,"type":"water","level":"basic"},
    "flood":{"name":"Flood","damage":20,"cooldown":4,"type":"water","level":"basic"},
    "light_beam":{"name":"Light Beam","damage":10,"cooldown":0,"type":"light","level":"alignment"},
    "solar_flare":{"name":"Solar Flare","damage":15,"cooldown":3,"type":"light","level":"alignment"},
    "radiant_burst":{"name":"Radiant Burst","damage":12,"cooldown":2,"type":"light","level":"alignment"},
    "photon_bolt":{"name":"Photon Bolt","damage":20,"cooldown":4,"type":"light","level":"alignment"},
    "void_strike":{"name":"Void Strike","damage":10,"cooldown":0,"type":"dark","level":"alignment"},
    "shadow_flux":{"name":"Shadow Flux","damage":12,"cooldown":2,"type":"dark","level":"alignment"},
    "nightmare":{"name":"Nightmare","damage":15,"cooldown":3,"type":"dark","level":"alignment"},
    "eclipse":{"name":"Eclipse","damage":20,"cooldown":5,"type":"dark","level":"alignment"},
    "space_rift":{"name":"Space Rift","damage":10,"cooldown":0,"type":"space","level":"cosmic"},
    "gravity_well":{"name":"Gravity Well","damage":15,"cooldown":3,"type":"space","level":"cosmic"},
    "space_wormhole":{"name":"Spatial Wormhole","damage":20,"cooldown":4,"type":"space","level":"cosmic"},
    "singularity":{"name":"Singularity","damage":25,"cooldown":5,"type":"space","level":"cosmic"},
    "galactic_strike":{"name":"Galactic Strike","damage":17,"cooldown":2,"type":"space","level":"cosmic"},
    "chronic_chakram":{"name":"Chronic Chakram","damage":10,"cooldown":0,"type":"time","level":"cosmic"},
    "temporal_loop":{"name":"Temporal Loop","damage":19,"cooldown":2,"type":"time","level":"cosmic"},
    "time_wormhole":{"name":"Temporal Wormhole","damage":20,"cooldown":4,"type":"time","level":"cosmic"},
    "fortune":{"name":"Fortune","damage":13,"cooldown":2,"type":"time","level":"cosmic"},
    "destiny":{"name":"Destiny","damage":16,"cooldown":2,"type":"time","level":"cosmic"}
}
def user_input(prompt,valid_options=None):
    user_is_stupid=True
    while user_is_stupid:
        try:
            response=input(f"{prompt} ({', '.join(valid_options[:-1])} or {valid_options[-1]}) ").strip() if valid_options else input(f"{prompt} ").strip()
            if valid_options and response not in valid_options:
                print(f"Invalid input. Please choose {', '.join(valid_options[:-1])} or {valid_options[-1]}.")
                continue
            user_is_stupid=False
            return response
        except Exception as e:
            print(f"Invalid input. Please try again. ({e})")
class stats:
    def __init__(self,attack,speed,defense,health):
        self.attack=attack
        self.speed=speed
        self.defense=defense
        self.health=health
class powers:
    def __init__(self,basic,alignment,cosmic):
        self.basic=basic
        self.alignment=alignment
        self.cosmic=cosmic
class player:
    def __init__(self):
        self.name=user_input("What would you like to name your character?")
        rand=user_input("Would you like to randomize your character's stats and powers?",["yes","no"]).lower()
        if rand=="yes":
            basic=random.choice(BASIC_POWERS*2+["nexus"])
            alignment=random.choice(ALIGNMENTS*4+["objectivity"])
            cosmic=random.choice(COSMIC_POWERS*5+["axiom"])
        else:
            basic=user_input("What would you like your character's basic power to be?",BASIC_POWERS).lower()
            alignment=user_input("What would you like your character's alignment to be?",ALIGNMENTS).lower()
            cosmic=user_input("What would you like your character's cosmic power to be?",COSMIC_POWERS).lower()
        self.stats=stats(10,20,10,100)
        self.powers=powers(basic,alignment,cosmic)
        self.known_moves=["flame_burst"if basic=="fire"else"iron_spike"if basic=="metal"else"splinter"if basic=="wood"else"pebble_shot"if basic=="earth"else"water_jet"]
        self.cooldown_times={}
        self.pets=[]
        self.rare_traits=((self.powers.basic=="nexus")+(self.powers.alignment=="objectivity")+(self.powers.cosmic=="axiom"))
class enemy:
    def __init__(self,name,stats,enemy_powers):
        self.name=name
        self.stats=stats
        self.powers=enemy_powers
        self.memory=powers("","","")
        self.known_moves=[]
        self.cooldown_times={}
def attack(attacker,attackee,move_used):
    attacker.cooldown_times[move_used]=MOVES[move_used]["cooldown"]
    attackee.stats.health-=MOVES[move_used]["damage"]
def battle_loop(player,enemy,turn):
    if turn==0:
        action="see moves"
        while action=="see moves":    
            action=user_input(f"You have {player.stats.attack} ATK, {player.stats.speed} SPD, {player.stats.defense} DFN, and {player.stats.health} HLT \nWhat would you like to do?",["use move","see moves"])
            if action=="use move":
                move_to_use=user_input("What move would you like to use?",[move for move in player.known_moves if move not in player.cooldown_times]) #how do i make it detect moves by name instead of id
                attack(player,enemy,move_to_use)
            elif action=="see moves":
                print("You can use the following moves: ")
                for move in player.known_moves[:-1]:
                    print(f"{MOVES[move]["name"]}: {MOVES[move]["damage"]} damage, {MOVES[move]["cooldown"]} cooldown")
                print(f"{MOVES[player.known_moves[-1]]["name"]}: {MOVES[player.known_moves[-1]]["damage"]} damage, {MOVES[player.known_moves[-1]]["cooldown"]} cooldown")
    for move in player.cooldown_times:
        player.cooldown_times[move]-=1
        if player.cooldown_times[move]<=0:
            del player.cooldown_times[move]
    for move in enemy.cooldown_times:
        enemy.cooldown_times[move]-=1
        if enemy.cooldown_times[move]<=0:
            del enemy.cooldown_times[move]
    return 0 if turn==1 else 1
print("Welcome to the game! You are a player in a world of magic and adventure. You will be able to choose your character's stats and powers, and then embark on a journey to defeat the evil forces that threaten the land.")
player=player()
print(f"""{player.name} has been created with the following stats: 
Attack: {player.stats.attack} 
Speed: {player.stats.speed} 
Defense: {player.stats.defense} 
Health: {player.stats.health} 
{'You are a Nexus!' if player.powers.basic=='nexus' else f'Your Basic Power: {player.powers.basic}'} 
{'You are an Objective!' if player.powers.alignment=='objectivity' else f'Your Alignment: {player.powers.alignment}'} 
{'You are an Axiom!' if player.powers.cosmic=='axiom' else f'Your Cosmic Power: {player.powers.cosmic}'}""")
if player.rare_traits==1:
    if player.powers.basic=='nexus':print("As a Nexus, you have the ability to control all elements and have access to powerful abilities. Use your powers wisely to defeat your enemies and protect the land ")
    if player.powers.alignment=='objectivity':print("As an Objective, you have the ability to balance the forces of light and dark and see the true nature of things. Use your powers wisely to maintain harmony in the land ")
    if player.powers.cosmic=='axiom':print("As an Axiom, you have the ability to manipulate space and time and have access to powerful abilities. Use your powers wisely to achieve your goals and explore the universe ")
elif player.rare_traits==2:
    if player.powers.basic=='nexus':
        if player.powers.alignment=='objectivity':print("As an Objective Nexus, you have the ability to control all elements and balance the forces of light and dark. Use your powers wisely to restore harmony in the land and defeat your enemies ")
        if player.powers.cosmic=='axiom':print("As an Axiomatic Nexus, you have the ability to manipulate space and time and have access to all the elements. Use your powers wisely to journey through the universe and defeat your enemies ")
    else:
        # If they have 2 rare traits, but not the Nexus trait, we can assume they have the other two rare traits
        print("As an Axiomatic Objective, you have the ability to manipulate space and time and balance the forces of light and dark. Use your powers wisely to restore harmony to the land and achieve your goals ")
elif player.rare_traits==3:
    print("Legend has it that one in a thousand people are Axiomatic Objective Nexum, and you are one of them! You have the ability to control all elements, manipulate space and time, and balance the forces of light and dark. Use your powers wisely to restore harmony to the land and achieve your goals ")
print("You are now ready to embark on your journey. Good luck, and may the forces of magic be with you!")
print("first enemy encounter testing an shi")
bob=enemy("Bob",stats(5,10,5,30),powers(random.choice(BASIC_POWERS),random.choice(ALIGNMENTS),random.choice(COSMIC_POWERS)))
battle_loop(player,bob,0)

File: magic.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>
#include <algorithm>
#include <random>
using std::cin;
using std::cout;
using std::string;
using std::vector;
using std::unordered_map;
std::random_device rd;
std::mt19937 gen(rd());
const vector<string> BASIC_POWERS={"fire","metal","wood","earth","water"};
const vector<string> ALIGNMENTS={"light","dark"};
const vector<string> COSMIC_POWERS={"space","time"};
const unordered_map<string,string>POWER_DEFINITIONS={
    {"fire","the power of flames and heat. You are able to control and manipulate flames, creating powerful attacks and defenses. You are also immune to fire and heat, allowing you to withstand extreme temperatures. You deal more damage to metal by melting it, but less to earth because it is not flammable. You have more defense towards wood since you can't put out fire by adding fuel, but less protection from water because water can extinguish fire."},
    {"metal","the power of strength and durability. You are able to control and manipulate metallic substanced, creating powerful weapons and armor. You are also immune to metal-based attacks such as bullets and blades, allowing you to withstand physical damage. You deal more damage to wood by cutting it, but less to water because water can cause rust. You have more defense towards earth since you are heavier than it, but less protection from fire because fire can melt metal."},
    {"wood","the power of growth and nature. You are able to control and manipulate plant life, creating powerful tools and structures. You are also immune to wood-based attacks, allowing you to withstand natural disasters. You deal more damage to earth since tree roots break up, penetrate, and bind soil together, but less to fire since you can't put out fire by adding fuel. You have more defense towards water because you drink water, but less protection from metal because metal can cut wood."},
    {"earth","the power of stability and protection. You are able to control and manipulate dirt, stone, and other earth materials, creating powerful barriers and fortifications. You are also immune to earth-based attacks, allowing you to withstand seismic activity. You deal more damage to water by soaking it up, but less to metal because metal is too durable. You have more defense towards fire you are non flammable, but less protection from wood because tree roots break up, penetrate, and bind soil together."},
    {"water","the power of fluidity and adaptability. You are able to control and manipulate water and ice, creating powerful waves and currents. You are also immune to water-based attacks, allowing you to withstand flooding. You deal more damage to fire by extinguishing it, but less to wood because wood can absorb water. You have more defense towards metal because it sinks in you, but less protection from earth because earth can absorb the water."},
    {"light","the power of illumination and vision. You are able to control and manipulate radiant energy, creating powerful beams and illusions. You are also immune to light-based attacks, allowing you to withstand bright environments. You deal more damage to dark by dispelling it, but take more damage from it as well since it obscures your vision."},
    {"dark","the power of negation and mystery. You are able to control and manipulate the absence of light and the shadows themself, creating powerful voids and illusions. You are also immune to dark-based attacks, allowing you to withstand eerie environments. You deal more damage to light by absorbing it, but take more damage to it as well because it blots out your shadows."},
    {"space","the power of dimensions and travel. You are able to control and manipulate the fabric of space, creating powerful portals and levitations. You are also immune to space-based attacks, allowing you to withstand vacuum and teleportation. You deal more damage to time by disrupting it, but take more damage from it as well since it is not in the 3 dimensions you exist in."},
    {"time","the power of past, present, and future. You are able to control and manipulate the flow of time, creating powerful time loops and see the future. You are also immune to time-based attacks, allowing you to withstand temporal anomalies. You deal more damage to space by disrupting it, but take more damage from it as well since it is not in your domain."}
};
struct Move {
    string name;
    int damage;
    int cooldown;
    string type;
    string level;
};
const unordered_map<string,Move>MOVES={
    {"flame_burst",{"Flame Burst",10,1,"fire","basic"}},
    {"fireball",{"Fireball",13,2,"fire","basic"}},
    {"heat_wave",{"Heat Wave",17,2,"fire","basic"}},
    {"iron_spike",{"Iron Spike",10,1,"metal","basic"}},
    {"shard_spray",{"Shard Spray",15,4,"metal","basic"}},
    {"knife_storm",{"Knife Storm",15,4,"metal","basic"}},
    {"splinter",{"Splinter",10,1,"wood","basic"}},
    {"vine_whip",{"Vine Whip",12,2,"wood","basic"}},
    {"leaf_storm",{"Leaf Storm",15,3,"wood","basic"}},
    {"pebble_shot",{"Pebble Shot",10,1,"earth","basic"}},
    {"tremor",{"Tremor",15,2,"earth","basic"}},
    {"boulder_crush",{"Boulder Crush",20,5,"earth","basic"}},
    {"water_jet",{"Water Jet",10,1,"water","basic"}},
    {"tidal_wave",{"Tidal Wave",15,2,"water","basic"}},
    {"flood",{"Flood",20,5,"water","basic"}},
    {"light_beam",{"Light Beam",10,1,"light","alignment"}},
    {"solar_flare",{"Solar Flare",15,4,"light","alignment"}},
    {"radiant_burst",{"Radiant Burst",12,3,"light","alignment"}},
    {"photon_bolt",{"Photon Bolt",20,5,"light","alignment"}},
    {"void_strike",{"Void Strike",10,1,"dark","alignment"}},
    {"shadow_flux",{"Shadow Flux",12,3,"dark","alignment"}},
    {"nightmare",{"Nightmare",15,4,"dark","alignment"}},
    {"eclipse",{"Eclipse",20,6,"dark","alignment"}},
    {"space_rift",{"Space Rift",10,1,"space","cosmic"}},
    {"gravity_well",{"Gravity Well",15,4,"space","cosmic"}},
    {"space_wormhole",{"Spatial Wormhole",20,5,"space","cosmic"}},
    {"singularity",{"Singularity",25,6,"space","cosmic"}},
    {"galactic_strike",{"Galactic Strike",17,3,"space","cosmic"}},
    {"chronic_chakram",{"Chronic Chakram",10,1,"time","cosmic"}},
    {"temporal_loop",{"Temporal Loop",19,3,"time","cosmic"}},
    {"time_wormhole",{"Temporal Wormhole",20,5,"time","cosmic"}},
    {"fortune",{"Fortune",13,3,"time","cosmic"}},
    {"destiny",{"Destiny",16,3,"time","cosmic"}}
};
string user_input(string prompt,vector<string>valid_options={}) {
    while(true){
        string response;
        cout<<prompt<<" ";
        if(!valid_options.empty()){
            cout<<"(";
            for(size_t i=0;i<valid_options.size()-1;i++) {
                cout<<valid_options[i];
                if(i<valid_options.size()-2) {
                    cout<<", ";
                }
            }
            cout<<" or "<<valid_options.back()<<") ";
        }
        cin>>response;
        if(!valid_options.empty()&&find(valid_options.begin(),valid_options.end(),response)==valid_options.end()){
            cout<<"Invalid input. Please choose ";
            for(size_t i=0;i<valid_options.size()-1;i++) {
                cout<<valid_options[i];
                if(i<valid_options.size()-2) {
                    cout<<", ";
                }
            }
            cout<<" or "<<valid_options.back()<<". ";
        }else{
            return(response);
        }
    }
}
struct Stats {
    int attack;
    int speed;
    int defense;
    int health;
};
struct Powers {
    string basic;
    string alignment;
    string cosmic;
};
struct Player {
    string name;
    int rare_traits=0;
    unordered_map<string,int>cooldown_times={};
    vector<string> known_moves;
    vector<string> pets;
    Stats stats{10,20,10,100};
    Powers powers;
    Player() {
        name=user_input("What would you like to name your character?");
        bool randomize=user_input("Would you like to randomize your character's stats and powers?",{"yes","no"})=="yes";
        if (randomize) {
            vector<string> basic_options={"fire","metal","wood","earth","water","fire","metal","wood","earth","water","nexus"};
            std::uniform_int_distribution<size_t> basic_dist(0,basic_options.size()-1);
            powers.basic=basic_options[basic_dist(gen)];
            vector<string> alignment_options={"light","light","light","light","dark","dark","dark","dark","objectivity"};
            std::uniform_int_distribution<size_t> alignment_dist(0,alignment_options.size()-1);
            powers.alignment=alignment_options[alignment_dist(gen)];
            vector<string> cosmic_options={"space","space","space","space","space","time","time","time","time","time","axiom"};
            std::uniform_int_distribution<size_t> cosmic_dist(0,cosmic_options.size()-1);
            powers.cosmic=cosmic_options[cosmic_dist(gen)];
        }else{
            powers.basic=user_input("What would you like your character's basic power to be?",BASIC_POWERS);
            powers.alignment=user_input("What would you like your character's alignment to be?",ALIGNMENTS);
            powers.cosmic=user_input("What would you like your character's cosmic power to be?",COSMIC_POWERS);
        }
        if (powers.basic=="fire"){known_moves.push_back("flame_burst");} 
        else if (powers.basic=="metal"){known_moves.push_back("iron_spike");} 
        else if (powers.basic=="wood"){known_moves.push_back("splinter");} 
        else if (powers.basic=="earth"){known_moves.push_back("pebble_shot");} 
        else if (powers.basic=="water"){known_moves.push_back("water_jet");} 
        else {vector<string> basic_moves={"flame_burst","iron_spike","splinter","pebble_shot","water_jet"};known_moves.insert(known_moves.end(),basic_moves.begin(),basic_moves.end());}
        if (powers.alignment=="light"){known_moves.push_back("light_beam");} 
        else if (powers.alignment=="dark"){known_moves.push_back("void_strike");} 
        else {vector<string> alignment_moves={"light_beam","void_strike"};known_moves.insert(known_moves.end(),alignment_moves.begin(),alignment_moves.end());}
        if (powers.cosmic=="space"){known_moves.push_back("space_rift");} 
        else if (powers.cosmic=="time"){known_moves.push_back("chronic_chakram");} 
        else {vector<string> cosmic_moves={"space_rift","chronic_chakram"};known_moves.insert(known_moves.end(),cosmic_moves.begin(),cosmic_moves.end());}
        if (powers.basic=="nexus"){rare_traits++;}
        if (powers.alignment=="objectivity"){rare_traits++;}
        if (powers.cosmic=="axiom"){rare_traits++;}
    }
};
int main() {
    Player player;
    return 0;
}